benchmarks/compare-spidermonkey-parsers.js

```diff
@@ -0,0 +1,259 @@
+    return performance.now() - start;
+}
+
+function for_all_files(parse, N = 1, prefix = "", result = {}) {
+    var path = "", content = "";
+    var t = 0;
+    var list = os.file.listDir(dir);
+    for (var file of list) {
+        try {
+            path = os.path.join(dir, file);
+            content = os.file.readRelativeToScript(path);
+            try {
+                t = 0;
+                for (var n = 0; n < N; n++)
+                    t += parse(content);
+                result[prefix + path] = { time: t / N, bytes: content.length };
+            } catch (e) {
+                // ignore all errors for now.
+                result[prefix + path] = { time: null, bytes: content.length };
+            }
+        } catch (e) {
+            // ignore all read errors.
+        }
+    }
+    return result;
+}
+
+function compare(res1, res2) {
+    var result = {
+        time1: 0,
+        time2: 0,
+        parsed_files: 0,
+        parsed_bytes: 0,
+        skipped_files: 0,
+        skipped_bytes: 0,
+        ratios_2over1: [],
+    };
+    for (var path of Object.keys(res1)) {
+        if (!(path in res1 && path in res2)) {
+            continue;
+        }
+        var p1 = res1[path];
+        var p2 = res2[path];
+        if (p1.time !== null && p2.time !== null) {
+            result.time1 += p1.time;
+            result.time2 += p2.time;
+            result.parsed_files += 1;
+            result.parsed_bytes += p1.bytes;
+            result.ratios_2over1.push(p2.time / p1.time);
+        } else {
+            result.skipped_files += 1;
+            result.skipped_bytes += p1.bytes;
+        }
+    }
+    return result;
+}
+
+function spread(table, min, max, step) {
+    // var chars = ["\xa0", "\u2591", "\u2592", "\u2593", "\u2588"];
+    var chars = ["\xa0", "\u2581", "\u2582", "\u2583", "\u2584", "\u2585", "\u2586", "\u2587", "\u2588"];
+    var s = ["\xa0", "\xa0", "" + min, "\xa0", "\xa0"];
+    var ending = ["\xa0", "\xa0", "" + max, "\xa0", "\xa0"];
+    var ranges = [];
+    var vmax = table.length / 10;
+    for (var i = min; i < max; i += step) {
+        ranges.push(0);
+    }
+    for (var x of table) {
+        if (x < min || max < x) continue;
+        var idx = ((x - min) / step)|0;
+        ranges[idx] += 1;
+    }
+    var max_index = chars.length * s.length;
+    var ratio = max_index / vmax;
+    for (i = 0; i < s.length; i++)
+        s[i] += "\xa0\u2595";
+    for (var v of ranges) {
+        var d = Math.min((v * ratio)|0, max_index - 1);
+        var offset = max_index;
+        for (i = 0; i < s.length; i++) {
+            offset -= chars.length;
+            var c = Math.max(0, Math.min(d - offset, chars.length - 1));
+            s[i] += chars[c];
+        }
+    }
+    for (i = 0; i < s.length; i++)
+        s[i] += "\u258f\xa0" + ending[i];
+    var res = "";
+    for (i = 0; i < s.length; i++)
+        res += "\n" + s[i];
+    return res;
+}
+
+function strategy_1() {
```

Comment:
Can you add documentation to those strategies?
like, what they're testing and what the differences are, and why only 0 is used?

Reply:
I will document in the patch as well, but answer here before.

What changes is the scheduling of the benchmarks, and this was made for testing if it matters or not in the reported results. Unfortunately, it matters.  Either because the code of the parser is hot in memory, or because the data being parsed is already in memory. This might have a non-neglectable impact on the reported performance.

Reply:
will it make sense to warm up and then run multiple times alternately?
or does it take too long?